### ISSUE TITLE
Bug with VW binary file while scoring.

### ISSUE THREAD

**Issue by gauravbansal2000** [JohnLangford/vowpal_wabbit#1392]
We trained our linear regressor using  : ` vw -d train.txt -f vw.model --holdout_off -c -k   --l1 6.98e-11 --passes 10 -b 25 -l 0.23 --power_t 0.34` and generated invert hash using `vw  -t train.txt  -i vw.model --invert_hash vw.readable.model.txt  --holdout_off    --l1 6.98e-11   -l 0.23 --power_t 0.34` .
We tried to score the same input in two different ways :- 

1. Running vw model in daemon mode.
2. Scoring manually using vw.readable.model.txt.

For the input "0_IN" 1st mode outputs 0 and by 2nd mode  we get a non zero number by adding the weight of "0_IN" and the Constant.
Command used to score in daemon mode :- `echo "| 0_IN " | netcat localhost 26541`

For the input "0_US" both scores from 2 modes match each other.
And if the input is `echo "| 0_IN 0_US " | netcat localhost 26541` even now both score from 2 modes match each other.

To summarize :- 

1. `echo "| 0_IN " | netcat localhost 26541` -> 0
2. `echo "| 0_US " | netcat localhost 26541` -> 0.551074
3. `echo "| 0_US 0_IN " | netcat localhost 26541` -> 0.241428
4. `echo "| 0_IN 0_US 0_IN " | nc localhost 26542` -> 0

Content in  vw.readable.model.txt :- 
0_IN:5448821:-0.309647
0_US:25727216:0.278704

Value of Constant = 0.27237

It seems from the 1st statement it is  finding  value for "0_IN" as -0.27237 and in 3rd statement  it is able to fetch the right weight of "0_IN" as -0.309647. And again in 4th statement weird things seem to happen. What might be the issue here?

**Comment by arielf** [comment]
Hi @gauravbansal2000 

I was not able to reproduce your example.  Got this error on the invert-hash step:
```
vw -t train.txt -i vw.model --invert_hash vw.readable.model.txt --holdout_off --l1 6.98e-11 -l 0.23 --power_t 0.34 
...
Reading datafile = train.txt
Error: need a cache file for multiple passes : try using --cache_file
```
The correct way to get the inverted-hash (readable model) is:
```
        vw -t train.txt -i vw.model --invert_hash vw.readable.model.txt
```
For the strict purpose of getting a readable model from an existing model, using the same data -- you don't need to specify multiple-passes and learning-rate options (`--l1, --l, --power_t` etc.)

More importantly:  what command-line exactly are you using for the daemon process?
Specifically: do you run it with `-t` (test only) to make sure the in-memory model is fixed, or not?  If you run without `-t` `vw` keeps updating the in-memory model, which I think is not what you want.

**Comment by gauravbansal2000** [comment]
Hi @arielf ,
We got this issue resolved. We were overlooking the parameters of the model over here.
This is first few lines of vw.readable.model.txt .
``` text
Version 8.3.1 
Id
Min label:0
Max label:8
bits:25
lda:0
0 ngram:
0 skip:
```
This model will clip the negative scores to zero because Min label is 0 over here. Problem in my case was "0_IN" feature  had a weight of -0.309647 , So the prediction for `| 0_IN` becomes 
``` text 
 -0.309647+(Constant=0.27237) = -0.037277
 ```
which then gets clipped to zero.

**Comment by arielf** [comment]
@gauravbansal2000 
Glad you figured this out.
Thanks for sharing what you learned!
